### PR TITLE
feat: refresh trending data from cmc intimately

### LINKED ISSUE #2276: [Demand] Refresh trending data from CMC intimately
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

The price data from the widget API is slow updating. The WebSocket price API (wss://stream.coinmarketcap.com/price/latest)  can subscribe the price of specific coin.

test

### PR DESCRIPTION
**Pull request by lelenei**
close #2276

### CODE REVIEW

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                    }
+                    client.send(JSON.stringify(params))
+                }
+                client.onmessage = (ev) => {
+                    if (typeof ev.data !== 'string') {
+                        return
+                    }
+                    try {
+                        const prices = JSON.parse(ev.data)
+                        resolve(prices['d']['cr'])
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts:44]: @Jack-Works async-rpc framework support non-standard json rpc field definition ?

Comment (Jack-Works) [reply packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts:44]: Nope, what does these fields stand for?

Comment (lelenei) [reply packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts]: this data is websocket's message, not async-rpc

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    v: number
+}
+
+let webSocket: WebSocket | null = null
+let timeout = 250
+let connectInterval: any = null
+
+function checkWebSocket(coinId: number) {
+    if (!webSocket || webSocket.readyState === WebSocket.CLOSED) {
+        openWebSocket(coinId)
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts:19]: https://cursed.im/retry

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    }
+}
+
+function openWebSocket(coinId: number): void {
+    if (!webSocket || webSocket.readyState === WebSocket.CLOSED) {
+        try {
+            const socket = new WebSocket(CMC_LATEST_BASE_URL)
+            socket.onopen = () => {
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts:27]:
```suggestion
            socket.addEventListener("open", () => {
```

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+function connectWebSocket() {
+    return new Promise<WebSocket>((resolve, reject) => {
+        try {
+            const client = new WebSocket(CMC_LATEST_BASE_URL)
+            resolve(client)
+        } catch (error) {
+            reject(error)
+        }
+    })
+}
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts:39]:
```suggestion
async function connectWebSocket() {
    return new WebSocket(CMC_LATEST_BASE_URL)
}
```

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
+import { CMC_LATEST_BASE_URL } from '../../constants/trending'
+import { EventEmitter } from 'events'
+
+const MAX_RETRY = 3
+let webSocket: WebSocket | null = null
+let retry = MAX_RETRY
+const Event = new EventEmitter()
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts:7]:
```suggestion
const events = new EventSource()
```

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                try {
+                    const jsondata = JSON.parse(ev.data)
+                    Event.emit(jsondata.id, jsondata)
+                } catch (e) {
+                    console.error(e)
+                }
+            }
+        })
+        .catch(error => console.log(error))
+    })
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts:97]:
```suggestion
    Event.addListener('subscribe', async (msg) => {
        if (webSocket && webSocket.readyState === WebSocket.OPEN) {
            if (WATCHED_COINS.some(x => x === msg)) {
                return
            }

            WATCHED_COINS.push(msg)
            webSocket.send(JSON.stringify(msgData))
            return
        }
       const client = connectWebSocket()
            client.onopen = () => {
                webSocket = client
                coinIds.map((x) => {
                    if (WATCHED_COINS.indexOf(x) === -1) {
                        WATCHED_COINS.push(x)
                    }
                })

                if (WATCHED_COINS.length > 0) {
                    client.send(JSON.stringify(msgData))
                }
            }
            client.onerror = (error) => {
                client.close()
                webSocket = null
                console.log(error)
            }
            client.onclose = () => {
                console.log('close')
                webSocket = null
            }

            client.onmessage = (ev) => {
                if (typeof ev.data!== 'string') {
                    return
                }
                try {
                    const jsondata = JSON.parse(ev.data)
                    Event.emit(jsondata.id, jsondata)
                } catch (e) {
                    console.error(e)
                }
            }
        }
    })
```

File: packages/maskbook/src/extension/options-page/DashboardRouters/Wallets.tsx [hunk 7]
```diff
@@ -25,7 +25,18 @@ import { WalletContent } from '../DashboardComponents/WalletContent'
 import { EthereumStatusBar } from '../../../web3/UI/EthereumStatusBar'
 import { extendsTheme } from '../../../utils/theme'
 import { useAssetsStableCoinDetailedDebank } from '../../../web3/hooks/useAssetsStableCoinDetailedDebank'
+import { getPrice } from '../../../plugins/Trader/apis/coinmarketcap'
+import { subscribeCoinPrce } from '../../../plugins/Trader/apis/coinmarketcap/getPriceStat'
 
+subscribeCoinPrce(1)
+subscribeCoinPrce(825)
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardRouters/Wallets.tsx:32]: what are those hardcoded magic number?

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts (hunk 4 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts:36]: and it doesn't have to be async

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts (hunk 5 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts:7]: No @septs EventSource is a completely different thing. https://developer.mozilla.org/en-US/docs/Web/API/EventSource

Do you mean `EventTarget`?

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/index.ts [hunk 8]
```diff
@@ -272,3 +273,10 @@ export async function getLatestMarketPairs(id: string, currency: string) {
     }
 }
 //#endregion
+
+//#region get price
+export function getPrice(coinId: number) {
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Trader/apis/coinmarketcap/index.ts:278]: export { getCoinPrice as getPrice } from '...'

File: packages/maskbook/src/extension/options-page/DashboardRouters/Wallets.tsx (hunk 7 above)
Comment (lelenei) [reply packages/maskbook/src/extension/options-page/DashboardRouters/Wallets.tsx:32]: this is test data

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/index.ts (hunk 8 above)
Comment (lelenei) [reply packages/maskbook/src/plugins/Trader/apis/coinmarketcap/index.ts:278]: this is debug code

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts (hunk 5 above)
Comment (septs) [reply packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts:7]: ye

**Comment by Jack-Works** [comment]
hi @lelenei please rebase this pr

File: packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+const Event = new EventEmitter()
+
+export interface PriceState {
+    id: number
+    mc: number
+    p: number
+    p1h: number
+    p7d: number
+    p24h: number
+    v: number
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Trader/apis/coinmarketcap/getPriceStat.ts:16]: would you like to add `/** */` style comment to notify people what thoes fields are

File: packages/maskbook/src/plugins/Trader/trending/useCurrentPrice.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import { useAsync } from 'react-use'
+import { isUndefined } from 'lodash-es'
+import { PluginTraderRPC } from '../messages'
+
+export function usePriceStat(coinId: number) {
+    return useAsync(async () => {
+        if (isUndefined(coinId)) return []
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Trader/trending/useCurrentPrice.ts:7]: I prefer `coidId === undefined`

File: packages/maskbook/src/plugins/Trader/trending/useCoinLatestPrice.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        latestPrice: number | null
+        latestPrice_1h_percentage: number | null
+    }>({
+        latestPrice: null,
+        latestPrice_1h_percentage: null,
+    })
+
+    const coins = useValueRef(subscribeCoins)
+
+    // @ts-ignore
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Trader/trending/useCoinLatestPrice.ts:18]: Why?

Comment (nuanyang233) [reply packages/maskbook/src/plugins/Trader/trending/useCoinLatestPrice.ts:18]: There is a `Not all code paths return a value` error, but I don't actually need return

Comment (Jack-Works) [reply packages/maskbook/src/plugins/Trader/trending/useCoinLatestPrice.ts:18]: You can return an empty function.  ts-ignore is too powerful and causes a refactoring hazard.

**Pull request closed without merging**